text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,111 @@
+
+Option 1 essentially reverts all transactions that happened after the block whose FTT has been exceeded. Option 2 attempts to revert the transactions from all forks but one. However, it is possible that some of the transactions on different forks of the chain don't conflict with each other. In such a case, we could attempt to replay all of them on the new trusted branch. This is the idea behind option 3: to gather the non-conflicting transactions, create a block containing them and collectively declare it correct, outside of the consensus algorithm.
+
+After the users collectively decide upon the last correct block, the transactions to be replayed and the list of equivocators to be slashed, the network can be restarted.
+
+### Restarting the network
+
+The process of restarting the network could be very similar to the process of starting it. In its essence, it is just like starting the network from scratch, but with a few differences:
+
+- There is a different set of genesis validators, which is based on the bonded validators of the last correct block, with the equivocators removed.
+- There is a different initial global state: it is not empty anymore, but it is the post-state of the last correct block.
+- Other details like the protocol version can differ as well, but they are not of the main concern for the restart process itself.
+
+The information necessary for restarting the network can (and probably should) be included in the chainspec. The "restart-chainspec" could be distributed among the validators the same way the initial chainspec for the network was.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Here we will just summarize the exact steps to be performed when an equivocation catastrophe happens.
+
+1. An equivocation catastrophe is detected. Once a single validator detects it, we can expect all the validators to detect it eventually as well (even if just due to them receiving the evidence from the validator that detected it first).
+2. Consensus gets suspended, though validators continue to gossip. The users (validator nodes' owners) receive notice about the catastrophe and start the process of social consensus.
+3. The users agree about the last correct block, the transactions that need to be replayed on top of it and the set of equivocating validators (exchanging evidence if necessary).
+4. A chainspec for the restart is manually constructed and agreed upon. This includes:
+    - the initial set of validators for the restarted network which excludes the equivocators,
+    - the initial global state hash for the restarted network
+5. The network is being restarted using the chainspec constructed above.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+The single drawback is that this proposal requires manual intervention in the case of an equivocation catastrophe. An ideal situation would be if the network could handle these automatically, but this is most probably impossible.
```

Comment:
Governance should tightly define the process to address an equivocation catastrophe, this reduces the amount that the process can be muddled with.

Reply:
I wouldn't say this is a drawback. That statement is a fact, and this CEP is proposed to remedy this issue in the first place.

Social consensus requires off-chain governance and manual intervention by default.

I'd say we have the following drawback: However we try to account for all possible scenarios with a legal process, there can always be one exception that this CEP has not accounted for. In that case, we would wait for it to play out, and eventually amend the CEP to account for the exception.